Rewrite LayoutTestFinder, making it own finding expectations

gsnedders opened the pull request:
#### bee9faa82faa4c52bc1c75dbc2f98b02c9d3f3a4
<pre>
Rewrite LayoutTestFinder, making it own finding expectations
<a href="https://bugs.webkit.org/show_bug.cgi?id=220421">https://bugs.webkit.org/show_bug.cgi?id=220421</a>
<a href="https://rdar.apple.com/problem/72894403">rdar://problem/72894403</a>

Reviewed by Jonathan Bedard.

Previously we have relied on the test runner to probe for expectation
files; this can end up taking a significant amount of time, as it
requires opening the directory file for each item in the baseline search
path, looking for one file at a time. Instead, we should just open each
directory once, when we are finding tests in that directory, and store
the paths to the expected results.

This refactoring also puts us in an easier place to implement
per-directory special-cases for LayoutTestFinder: this is most obviously
useful for making http/wpt and imported/w3c/web-platform-tests more
closely match WPT semantics (and a future patch will move to primarily
relying on the WPT manifest to find tests in these directories, rather
than our existing behaviour).

* Tools/Scripts/webkitpy/common/system/filesystem.py:
(FileSystem):
(FileSystem.dirs_under):
(FileSystem.dirs_under.filter_all):
(FileSystem.files_under):
(FileSystem.islink):
(FileSystem.scandir):
(FileSystem.mkdtemp):
* Tools/Scripts/webkitpy/common/system/filesystem_mock.py:
(MockFileSystem.islink):
(MockFileSystem.scandir):
(ScanDirMock):
(ScanDirMock.__init__):
(ScanDirMock.__iter__):
(ScanDirMock.__enter__):
(ScanDirMock.__exit__):
(ScanDirMock.close):
(DirEntryMock):
(DirEntryMock.__init__):
(DirEntryMock.name):
(DirEntryMock.is_dir):
(DirEntryMock.is_file):
(DirEntryMock.is_symlink):
(DirEntryMock.stat):
(DirEntryMock.inode):
* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder.py: Added.
(natsort):
(LayoutTestFinder):
(LayoutTestFinder.__init__):
(LayoutTestFinder._load_w3c_resource_data):
(LayoutTestFinder._canonicalize_test_path):
(LayoutTestFinder.get_tests):
(LayoutTestFinder._split_glob):
(LayoutTestFinder._get_tests_in_directory):
(LayoutTestFinder._process_directory):
(LayoutTestFinder.is_test_file):
(LayoutTestFinder._tests_for_path):
(LayoutTestFinder._find_variants):
(LayoutTestFinder._percent_encoded_variant):
(LayoutTestFinder._is_valid_variant):
(LayoutTestFinder._expectations_for_test):
(LayoutTestFinder.is_wpt_crash_test):
* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy.py:
(_is_reference_html_file):
(LayoutTestFinder.__init__):
(LayoutTestFinder.find_tests_by_path):
(LayoutTestFinder._is_test_file):
(LayoutTestFinder._is_w3c_resource_file):
(_has_supported_extension): Deleted.
(LayoutTestFinder._is_wpt_test): Deleted.
(LayoutTestFinder._is_wpt_crash_test): Deleted.
(LayoutTestFinder._expand_variants): Deleted.
(LayoutTestFinder._percent_encoded_variant): Deleted.
(LayoutTestFinder._is_valid_variant): Deleted.
(LayoutTestFinder._real_tests): Deleted.
* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy_unittest.py:
(LayoutTestFinderTests.tearDown):
(LayoutTestFinderTests.test_find_glob_b2):
(LayoutTestFinderTests.test_supported_test_extensions): Deleted.
* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_unittest.py: Added.
(LayoutTestFinderTests):
(LayoutTestFinderTests.__init__):
(LayoutTestFinderTests.setUp):
(LayoutTestFinderTests.tearDown):
(LayoutTestFinderTests.test_split_glob):
* Tools/Scripts/webkitpy/layout_tests/run_webkit_tests.py:
(_print_expectations):
* Tools/Scripts/webkitpy/layout_tests/run_webkit_tests_integrationtest.py:
(RunTest.test_device_type_specific_listing):
(RunTest.test_ipad_listing):
* Tools/Scripts/webkitpy/port/test.py:
* Tools/Scripts/webkitpy/test/main.py:
(main):
* Tools/Scripts/webkitpy/w3c/test_importer_unittest.py:
(TestImporterTest): Deleted.
(TestImporterTest._parse_options): Deleted.
(TestImporterTest.test_import_dir_with_no_tests_and_no_hg): Deleted.
(TestImporterTest.test_import_dir_with_no_tests): Deleted.
(TestImporterTest.test_import_dir_with_empty_init_py): Deleted.
(TestImporterTest.import_directory): Deleted.
(TestImporterTest.import_downloaded_tests): Deleted.
(TestImporterTest.test_harnesslinks_conversion): Deleted.
(TestImporterTest.test_skip_test_import_download): Deleted.

Canonical link: <a href="https://commits.webkit.org/276129@main">https://commits.webkit.org/276129@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b61eed66c9c283e116ebd36212dc7677dd42be91

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43793 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22842 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46230 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46436 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39891 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46097 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26723 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20249 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36162 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44367 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19882 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37724 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17154 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/43660 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17411 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38797 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1847 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39974 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39100 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47993 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18820 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15403 "Found 1 new test failure: http/tests/navigation/parsed-iframe-dynamic-form-back-entry.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42962 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20223 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41652 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9745 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20419 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19847 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/webkitpy/common/system/filesystem.py:
```diff
@@ -134,7 +144,12 @@ def filter_all(fs, dirpath, basename):
             return []
 
-        for (dirpath, dirnames, filenames) in os.walk(path):
+        if scandir is not None:
```
Why aren't we unconditionally using `scandir`?

gsnedders replied:
Oh, I should add a comment there. We could just make this `if sys.version_info < (3,)`.

In short: the Python 2 package isn't overwritten by `pyfakefs`, and it directly makes calls (from C) that interact with the filesystem that `pyfakefs` cannot override.

Fundamentally, I don't think it's worthwhile caring about the perf cost of always falling back to `os.walk` on Python 2 at this point.

JonWBedard replied:
I think making this `sys.version_info < (3,)` would be better, so we know we can rip it out after we deprecate Python 2.

JonWBedard commented on Tools/Scripts/webkitpy/conftest.py:
```diff
@@ -110,3 +110,6 @@ def pytest_collection_modifyitems(config, items):
     AutoInstall.install_everything()
+
+
+pytest.register_assert_rewrite("webkitpy.layout_tests.controllers.layout_test_finder_legacy")
```
What is the purpose of this line?

gsnedders replied:
this is in the "OOPS" commit; it turns on assertion rewriting in pytest (c.f. https://docs.pytest.org/en/latest/how-to/assert.html#assertion-introspection-details) for that module, so pytest gives useful diffs when `assert new_tests == old_tests` fails

gsnedders commented on Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder.py:
```diff
@@ -0,0 +1,614 @@
+
+        # XXX: both this and the realpath calls above mean __init__ reads the
+        # filesystem, and object data reflects FS state this time; are we okay with
+        # this?
```
This API design question I think is the only real outstanding question on this PR.

JonWBedard replied:
This seems reasonable, I struggle to think how we would structure this except to capture filesystem state on invocation.

JonWBedard commented on Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_unittest.py:
```diff
@@ -0,0 +1,131 @@
+        self.assertFalse(finder.is_test_file("", "foo-ref.xhtml"))
+        self.assertFalse(finder.is_test_file("", "ref-foo.html"))
+        self.assertFalse(finder.is_test_file("", "notref-foo.xhr"))
+        self.assertFalse(finder.is_test_file("", "foo_wsh.py"))
```
We have some Python CGI scripts, right? That means `is_test_file` isn't quite doing what I'd expect.

gsnedders replied:
https://github.com/WebKit/WebKit/blob/3412f17868274cc523f247c8dbc336a73d273abe/Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy.py#L48-L50 — this is a special-case on `main` today

JonWBedard replied:
Should we have a `.py` file in here to make it clear that `.py` files _are_ a valid test name?

JonWBedard commented on Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder.py:
```diff
@@ -0,0 +1,614 @@
+
+def natsort(string_to_split):
+    split = digit_re.split(string_to_split)
+    # this converts the split numbers into tuples so that "01" < "1"
```
Probably should be `"1" < "02"` as your example, right?

gsnedders replied:
No; the goal here is to provide a strict total ordering of strings, which means we need to be able to distinguish "01" and "1".

JonWBedard commented on Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder.py:
```diff
@@ -0,0 +1,614 @@
+        path = fs.realpath(path)
+
+        relpath = fs.relpath(path, base)
+        if relpath.startswith(".." + fs.sep):
```
Is this the only case we want to convert to an `abspath`?

gsnedders replied:
I think so? Can you think of any other?

gsnedders committed:
Rewrite LayoutTestFinder, making it own finding expectations
https://bugs.webkit.org/show_bug.cgi?id=220421
rdar://problem/72894403

Reviewed by Jonathan Bedard.

Previously we have relied on the test runner to probe for expectation
files; this can end up taking a significant amount of time, as it
requires opening the directory file for each item in the baseline search
path, looking for one file at a time. Instead, we should just open each
directory once, when we are finding tests in that directory, and store
the paths to the expected results.

This refactoring also puts us in an easier place to implement
per-directory special-cases for LayoutTestFinder: this is most obviously
useful for making http/wpt and imported/w3c/web-platform-tests more
closely match WPT semantics (and a future patch will move to primarily
relying on the WPT manifest to find tests in these directories, rather
than our existing behaviour).

* Tools/Scripts/webkitpy/common/system/filesystem.py:
(FileSystem):
(FileSystem.dirs_under):
(FileSystem.dirs_under.filter_all):
(FileSystem.files_under):
(FileSystem.islink):
(FileSystem.scandir):
(FileSystem.mkdtemp):
* Tools/Scripts/webkitpy/common/system/filesystem_mock.py:
(MockFileSystem.islink):
(MockFileSystem.scandir):
(ScanDirMock):
(ScanDirMock.__init__):
(ScanDirMock.__iter__):
(ScanDirMock.__enter__):
(ScanDirMock.__exit__):
(ScanDirMock.close):
(DirEntryMock):
(DirEntryMock.__init__):
(DirEntryMock.name):
(DirEntryMock.is_dir):
(DirEntryMock.is_file):
(DirEntryMock.is_symlink):
(DirEntryMock.stat):
(DirEntryMock.inode):
* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder.py: Added.
(natsort):
(LayoutTestFinder):
(LayoutTestFinder.__init__):
(LayoutTestFinder._load_w3c_resource_data):
(LayoutTestFinder._canonicalize_test_path):
(LayoutTestFinder.get_tests):
(LayoutTestFinder._split_glob):
(LayoutTestFinder._get_tests_in_directory):
(LayoutTestFinder._process_directory):
(LayoutTestFinder.is_test_file):
(LayoutTestFinder._tests_for_path):
(LayoutTestFinder._find_variants):
(LayoutTestFinder._percent_encoded_variant):
(LayoutTestFinder._is_valid_variant):
(LayoutTestFinder._expectations_for_test):
(LayoutTestFinder.is_wpt_crash_test):
* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy.py:
(_is_reference_html_file):
(LayoutTestFinder.__init__):
(LayoutTestFinder.find_tests_by_path):
(LayoutTestFinder._is_test_file):
(LayoutTestFinder._is_w3c_resource_file):
(_has_supported_extension): Deleted.
(LayoutTestFinder._is_wpt_test): Deleted.
(LayoutTestFinder._is_wpt_crash_test): Deleted.
(LayoutTestFinder._expand_variants): Deleted.
(LayoutTestFinder._percent_encoded_variant): Deleted.
(LayoutTestFinder._is_valid_variant): Deleted.
(LayoutTestFinder._real_tests): Deleted.
* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy_unittest.py:
(LayoutTestFinderTests.tearDown):
(LayoutTestFinderTests.test_find_glob_b2):
(LayoutTestFinderTests.test_supported_test_extensions): Deleted.
* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_unittest.py: Added.
(LayoutTestFinderTests):
(LayoutTestFinderTests.__init__):
(LayoutTestFinderTests.setUp):
(LayoutTestFinderTests.tearDown):
(LayoutTestFinderTests.test_split_glob):
* Tools/Scripts/webkitpy/layout_tests/run_webkit_tests.py:
(_print_expectations):
* Tools/Scripts/webkitpy/layout_tests/run_webkit_tests_integrationtest.py:
(RunTest.test_device_type_specific_listing):
(RunTest.test_ipad_listing):
* Tools/Scripts/webkitpy/port/test.py:
* Tools/Scripts/webkitpy/test/main.py:
(main):
* Tools/Scripts/webkitpy/w3c/test_importer_unittest.py:
(TestImporterTest): Deleted.
(TestImporterTest._parse_options): Deleted.
(TestImporterTest.test_import_dir_with_no_tests_and_no_hg): Deleted.
(TestImporterTest.test_import_dir_with_no_tests): Deleted.
(TestImporterTest.test_import_dir_with_empty_init_py): Deleted.
(TestImporterTest.import_directory): Deleted.
(TestImporterTest.import_downloaded_tests): Deleted.
(TestImporterTest.test_harnesslinks_conversion): Deleted.
(TestImporterTest.test_skip_test_import_download): Deleted.

Canonical link: https://commits.webkit.org/276129@main
